astropy/units/quantity_helper.py

```diff
@@ -307,40 +307,51 @@ def helper_two_arg_dimensionless(f, unit1, unit2):
 UFUNC_HELPERS[np.logaddexp2] = helper_two_arg_dimensionless
 
 
-def get_converters_and_unit(f, *units):
-
+def get_converters_and_unit(f, unit1, unit2):
     converters = [None, None]
-    # no units for any input -- e.g., np.add(a1, a2, out=q)
-    if all(unit is None for unit in units):
-        return converters, dimensionless_unscaled
+    # By default, we try adjusting unit2 to unit1, so that the result will
+    # be unit1 as well. But if there is no second unit, we have to try
+    # adjusting unit1 (to dimensionless, see below).
+    if unit2 is None:
+        if unit1 is None:
+            # No units for any input -- e.g., np.add(a1, a2, out=q)
+            return converters, dimensionless_unscaled
+        changeable = 0
+        # swap units.
+        unit2 = unit1
+        unit1 = None
+    elif unit2 is unit1:
+        # ensure identical units is fast ("==" is slow, so avoid that).
+        return converters, unit1
+    else:
+        changeable = 1
 
-    fixed, changeable = (1, 0) if units[1] is None else (0, 1)
-    if units[fixed] is None:
+    # Try to get a converter from unit2 to unit1.
+    if unit1 is None:
         try:
-            converters[changeable] = get_converter(units[changeable],
+            converters[changeable] = get_converter(unit2,
                                                    dimensionless_unscaled)
         except UnitsError:
             # special case: would be OK if unitless number is zero, inf, nan
-            converters[fixed] = False
-            return converters, units[changeable]
+            converters[1-changeable] = False
+            return converters, unit2
         else:
             return converters, dimensionless_unscaled
 
     else:
         try:
-            converters[changeable] = get_converter(units[changeable],
-                                                   units[fixed])
+            converters[changeable] = get_converter(unit2, unit1)
         except UnitsError:
             raise UnitConversionError(
                 "Can only apply '{0}' function to quantities "
                 "with compatible dimensions"
                 .format(f.__name__))
 
-        return converters, units[fixed]
+        return converters, unit1
 
 
-def helper_twoarg_invariant(f, unit1, unit2):
-    return get_converters_and_unit(f, unit1, unit2)
+# Use function directly so to avoid a needless function call.
+helper_twoarg_invariant = get_converters_and_unit
```

Comment:
Is `helper_twoarg_invariant` in the public API or can we just replace all? If we can't, can you leave a comment to say why we didn't replace all?

Reply:
It is not directly public, but all the helper functions are called this way, so I felt it was slightly cleaner. I'll add a comment to that extent.